src/Framework/Constraint/Count.php

```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
             $this->expectedCount
         );
     }
+
+    private function getCountOfTraversable(Traversable $traversable): int
+    {
+        if ($this->traversableCounts === null) {
+            $this->traversableCounts = new \SplObjectStorage();
+        }
+
+        while ($traversable instanceof IteratorAggregate) {
+            $traversable = $traversable->getIterator();
+        }
+
+        if ($this->traversableCounts->contains($traversable)) {
+            return $this->traversableCounts[$traversable];
+        }
+
+        if ($traversable instanceof Iterator) {
+            return $this->countIteratorWithoutChangingPositionIfPossible($traversable);
+        }
+
+        return $this->getCountOfTraversableThatIsNotIteratorOrIteratorAggregate($traversable);
+    }
+
+    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
+    {
+        $key   = $iterator->key();
+        $count = $this->getCountOfIterator($iterator);
+
+        $this->traversableCounts->attach($iterator, $count);
+
+        if ($this->iteratorIsRewindable($iterator)) {
+            $this->traversableCounts->detach($iterator);
+            $this->rewindIterator($iterator, $key);
```

Comment:
This originally was inside a condition `if ($key !== null) {`

Reply:
Many changes where made compared to the original. If you put in this condition now the tests will fail.

Reply:
(If you want to see how it got removed, there are plenty of small commits to look at)